lib/client.js

```diff
@@ -1404,150 +1407,276 @@ function write (client, request) {
+      .removeListener('error', onFinished)
+    body
+      .removeListener('data', onData)
+      .removeListener('end', onFinished)
+      .removeListener('error', onFinished)
+      .removeListener('close', onAbort)
+
+    // TODO (fix): Avoid using err.message for logic.
+    if (err && (err.code !== 'UND_ERR_INFO' || err.message !== 'reset')) {
+      util.destroy(body, err)
+    } else {
+      util.destroy(body)
+    }
+
+    if (err) {
+      assert(client[kRunning] <= 1, 'pipeline should only contain this request')
+      util.destroy(socket, err)
+    }
 
-      socket
-        .removeListener('drain', onDrain)
-        .removeListener('error', onFinished)
-      body
-        .removeListener('data', onData)
-        .removeListener('end', onFinished)
-        .removeListener('error', onFinished)
-        .removeListener('close', onAbort)
-
-      // TODO (fix): Avoid using err.message for logic.
-      if (err && (err.code !== 'UND_ERR_INFO' || err.message !== 'reset')) {
-        util.destroy(body, err)
+    if (socket.destroyed) {
+      return
+    }
+
+    if (bytesWritten === 0) {
+      if (expectsPayload) {
+        // https://tools.ietf.org/html/rfc7230#section-3.3.2
+        // A user agent SHOULD send a Content-Length in a request message when
+        // no Transfer-Encoding is sent and the request method defines a meaning
+        // for an enclosed payload body.
+
+        socket.write(`${header}content-length: 0\r\n\r\n\r\n`, 'ascii')
       } else {
-        util.destroy(body)
+        socket.write(`${header}\r\n`, 'ascii')
+      }
+    } else if (contentLength === null) {
+      socket.write('\r\n0\r\n\r\n', 'ascii')
+    }
+
+    // TODO (fix): Add comment clarifying what this does?
+    if (socket[kParser].timeout && socket[kParser].timeoutType === TIMEOUT_HEADERS) {
+      // istanbul ignore else: only for jest
+      if (socket[kParser].timeout.refresh) {
+        socket[kParser].timeout.refresh()
+      }
+    }
+
+    resume(client)
+  }
+
+  body
+    .on('data', onData)
+    .on('end', onFinished)
+    .on('error', onFinished)
+    .on('close', onAbort)
+
+  socket
+    .on('drain', onDrain)
+    .on('error', onFinished)
+}
+
+async function writeIterable ({ client, request, socket, contentLength, header, expectsPayload }) {
+  const { body } = request
+
+  assert(contentLength !== 0 || client[kRunning] === 0, 'iterator body cannot be pipelined')
+
+  let callback = noop
+  function onDrain () {
+    const cb = callback
+    callback = noop
+    cb()
+  }
+
+  const waitForDrain = () => new Promise((resolve) => {
+    assert(callback === noop)
+    callback = resolve
+  })
+
+  socket[kWriting] = true
+  socket
+    .on('close', onDrain)
+    .on('drain', onDrain)
+
+  let bytesWritten = 0
+  try {
+    // TODO (fix): What if socket errors while waiting for body?
+    // It's up to the user to somehow abort the async iterable.
+    for await (const chunk of body) {
+      if (socket[kError]) {
```

Comment:
Is this enough? Could the socket somehow get closed without setting kError?

Reply:
No, we ensure it always has error